src/storage/page_buffer.c

```diff
@@ -8319,35 +8332,46 @@ pgbuf_read_page_from_file (THREAD_ENTRY * thread_p, const VPID * vpid, void *io_
 /*
  * pgbuf_request_data_page_from_page_server () - Sends a request for a page to Page Server.
  *   return: void
+ *
+ * vpid(in) : page identifier
+ * fetch_mode(in) : fetch mode
+ * target_repl_lsa(int) : an LSA sent to the Page Server which the Page Server's replication must have
+ *                        advanced past; part of the replication mechanism
+ * io_page(out) : Address where content of page is stored. Must be of sufficient size.
  */
 static int
-pgbuf_request_data_page_from_page_server (const VPID * vpid, log_lsa target_repl_lsa, FILEIO_PAGE * io_page)
+pgbuf_request_data_page_from_page_server (const VPID * vpid, const PAGE_FETCH_MODE fetch_mode,
+					  log_lsa target_repl_lsa, FILEIO_PAGE * io_page)
 {
   assert (get_server_type () == SERVER_TYPE_TRANSACTION);
 
-  // *INDENT-OFF*
-  /* Send a request to Page Server for the Page. */
-
   cubpacking::packer pac;
   size_t size = 0;
 
   size += cublog::lsa_utils::get_packed_size (pac, size);
   size += vpid_utils::get_packed_size (pac, size);
   size += pac.get_packed_int_size (size);
-  std::unique_ptr<char[]> buffer (new char[size]);
+  std::unique_ptr < char[] > buffer (new char[size]);
 
   pac.set_buffer (buffer.get (), size);
   vpid_utils::pack (pac, *vpid);
   cublog::lsa_utils::pack (pac, target_repl_lsa);
 
-  PAGE_FETCH_MODE fetch_mode;
+  // register a fatal error if an unexpected fetch mode
   if (log_is_in_crash_recovery_and_not_yet_completes_redo ())
     {
-      fetch_mode = RECOVERY_PAGE;
+      if (!(fetch_mode == RECOVERY_PAGE || fetch_mode == OLD_PAGE || fetch_mode == OLD_PAGE_PREVENT_DEALLOC))
+	{
+	  er_set (ER_FATAL_ERROR_SEVERITY, ARG_FILE_LINE, ER_GENERIC_ERROR, 0);
+	}
     }
-    else
+  else
     {
-      fetch_mode = OLD_PAGE;
+      if (!(fetch_mode == OLD_PAGE || fetch_mode == OLD_PAGE_MAYBE_DEALLOCATED
```

Comment:
When the replicator fixes pages so that it retrieves from PS, isn't it supposed to use OLD_PAGE_IF_IN_BUFFER_OR_IN_TRANSIT?

Reply:
I tried it on your branch, but it fails because of an assertion fails: 
```
cub_server: /home/cubrid/cubrid/lets/src/transaction/log_replication_atomic.cpp:109: virtual void cublog::atomic_replicator::redo_upto(cubthread::entry&, const log_lsa&): Assertion `!m_atomic_helper.is_part_of_atomic_replication (header.trid)' failed.
```
So, just asking :).